src/prof_sys_linux.c

```diff
@@ -0,0 +1,153 @@
+  //
+  // The fields we care about are always within the first 34 characters so
+  // as long as `buf` contains the start of a mapping line it can always be
+  // parsed.
+  static const int kMappingFieldsWidth = 34;
+
+  int fd = -1;
+  char buf[4096];
+  ssize_t remaining = 0; // actual number of bytes read to buf
+  char* line = NULL;
+
+  while (1) {
+    if (fd < 0) {
+      // case 0: initial open of maps file
+      fd = malloc_open(maps_path, O_RDONLY);
+      if (fd < 0) {
+        return errno;
+      }
+
+      remaining = malloc_read_fd(fd, buf, sizeof(buf));
+      if (remaining <= 0) {
+        break;
+      }
+      line = buf;
+    } else if (line == NULL) {
+      // case 1: no newline found in buf
+      remaining = malloc_read_fd(fd, buf, sizeof(buf));
+      if (remaining <= 0) {
+        break;
+      }
+      line = memchr(buf, '\n', remaining);
+      if (line != NULL) {
+        line++; // advance to character after newline
+        remaining -= (line - buf);
+      }
+    } else if (line != NULL && remaining < kMappingFieldsWidth) {
+      // case 2: found newline but insufficient characters remaining in buf
+
+      // fd currently points to the character immediately after the last
+      // character in buf. Seek fd to the character after the newline.
+      if (malloc_lseek(fd, -remaining, SEEK_CUR) == -1) {
+        ret = errno;
+        break;
+      }
+
+      remaining = malloc_read_fd(fd, buf, sizeof(buf));
+      if (remaining <= 0) {
+        break;
+      }
+      line = buf;
+    } else {
+      // case 3: found newline and sufficient characters to parse
+
+      // parse <start>-<end>
+      char* tmp = line;
+      uintptr_t start_addr = strtoul(tmp, &tmp, 16);
+      if (addr >= start_addr) {
+        tmp++; // advance to character after '-'
+        uintptr_t end_addr = strtoul(tmp, &tmp, 16);
+        if (addr < end_addr) {
+          *mm_start = start_addr;
+          *mm_end = end_addr;
+          ret = 0;
+          break;
+        }
+      }
+
+      // Advance to character after next newline in the current buf.
+      char* prev_line = line;
+      line = memchr(line, '\n', remaining);
+      if (line != NULL) {
+        line++; // advance to character after newline
+        remaining -= (line - prev_line);
+      }
+    }
+  }
+
+  malloc_close(fd);
+  return ret;
+}
+
+static uintptr_t prof_main_thread_stack_start(const char* stat_path) {
+  uintptr_t stack_start = 0;
+
+  int fd = malloc_open(stat_path, O_RDONLY);
+  if (fd < 0) {
+    return 0;
+  }
+
+  char buf[512];
+  ssize_t n = malloc_read_fd(fd, buf, sizeof(buf) - 1);
+  if (n >= 0) {
+    buf[n] = '\0';
+    if (sscanf(
+            buf,
+            "%*d (%*[^)]) %*c %*d %*d %*d %*d %*d %*u %*u %*u %*u %*u %*u %*u %*d %*d %*d %*d %*d %*d %*u %*u %*d %*u %*u %*u %ld",
+            &stack_start) != 1) {
+    }
+  }
+  malloc_close(fd);
+  return stack_start;
+}
+
+uintptr_t prof_thread_stack_start(uintptr_t stack_end) {
```

Comment:
Question: is it true that this is the part that makes it Linux specific? If other platforms would like to utilize the framepointer based unwinder, as long as they implement this part, the rest of the mechanism would work?

Reply:
The determination of stack address range - yes, but also the implementation of  `prof_backtrace_impl()` is making assumptions about stack layout and growth direction that are not universally true - though quite common (the gcc intrinsics hide the details of that). 

It might be possible to fold the address checks into the existing `gcc` intrinsics unwinder - they would still only be available for linux (for now) of course - but it might be easier to add support for new platforms.

Reply:
Actually - looking at the generated code for gcc intrinsics based `prof_backtrace_impl()` we probably should avoid that where possible.

# Codegen

The gcc intrinsics approach uses `__builtin_frame_address()` to sanity check the frame - along with `__builtin_return_address()` to get the actual return address that is the unwound stack frame. The problem is that for each of those the compiler (Clang) doesn't seem to realize the relationship between `BT_FRAME(N)` and `BT_FRAME(N+1)` and the generated code becomes O(N^2).

## Clang (trunk)
https://godbolt.org/z/cs7sjPchn

Clang is re-walking the stack twice on every `BT_FRAME(N)`:

`BT_FRAME(4)`
```
        mov     rax, qword ptr [rbp]       # __builtin_frame_address()
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        cmp     qword ptr [rax], 0
        je      .LBB0_31
        mov     rax, qword ptr [rbp]     # __builtin_return_address()
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax + 8]
        test    rax, rax
        je      .LBB0_31
        mov     qword ptr [rdi + 32], rax
        mov     dword ptr [rsi], 5
        cmp     edx, 5
        je      .LBB0_31
```

`BT_FRAME(8)`
```
        mov     rax, qword ptr [rbp]       # __builtin_frame_address()
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        cmp     qword ptr [rax], 0
        je      .LBB0_31
        mov     rax, qword ptr [rbp]     # __builtin_return_address()
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax]
        mov     rax, qword ptr [rax + 8]
        test    rax, rax
        je      .LBB0_31
        mov     qword ptr [rdi + 64], rax
        mov     dword ptr [rsi], 9
        cmp     edx, 9
        je      .LBB0_31
```

## GCC (trunk)
https://godbolt.org/z/KoYEhsx7o

GCC does better - it appears to at least partially combines subsequent instances of `BT_FRAME(N)` but is still re-walking the stack each iteration (but not twice like Clang).

# Benchmarks

Testing with a small benchmark that unwinds 40 frame stacks in a loop:

libunwind: 284.00us
gcc: 404.59us
fp-linux: 222.82us

The benchmark bears out the inefficient codegen with the gcc intrinsics path (this was using Clang which is our default compiler).

Reply:
Back to your original question - is it possible to make the `fp-linux` unwinder work for other platforms - I think so.  It would be better to have a less Linux-specific naming in that case.